src/persist-client/src/read.rs

```diff
@@ -881,34 +888,49 @@ where
         self.explicitly_expired = true;
     }
 
-    async fn next_listen_batch(&mut self, frontier: &Antichain<T>) -> HollowBatch<T> {
+    async fn next_listen_batch(
+        &mut self,
+        frontier: &Antichain<T>,
+        watch: &mut StateWatch<K, V, T, D>,
+    ) -> HollowBatch<T> {
         let mut retry: Option<MetricsRetryStream> = None;
         loop {
-            if let Some(b) = self.machine.next_listen_batch(frontier) {
-                return b;
-            }
+            let seqno = match self.machine.next_listen_batch(frontier) {
+                Ok(b) => return b,
+                Err(seqno) => seqno,
+            };
             // Only sleep after the first fetch, because the first time through
```

Comment:
oops, missed this before -- think we'll want to update this to be "after the first fetch or after a state notification" since the latter resets our retry

Reply:
actually, I'm now undecided on whether `wait_for_seqno_ge` returning first should reset the retry interval... seems like there could be some odd interactions:

* listening for a frontier in the future
* retry interval starts to back off
* unrelated seqno bump happens (e.g. for a heartbeat)
* retryer resets to fast polling

Reply:
lets talk this over offline at some point

Reply:
okay, fixed this and also gave the watch treatment to snapshot

this required a pretty fundamental rewrite of the loop(s), now to be in terms of a FuturesUnordered that always has two things in it. the new versions seem more intimidating, but a large part of that is the boilerplate of the enum + box pin dyn (possible there's some nice cleanup here that would help?). i think they're actually much easier to read and reason about than the old loops, and they'll also be easier to make changes to

one minor functional difference is that the snapshot loop now bases the info vs debug log on how long it's been looking for a snapshot _total_, not how long the next sleep will be... but that's what I meant to do in the first place XD

I ran out of time today to test this more than getting existing persist unit tests passing, so totally possible this fails every CI job, but figured i'd get the new version out for your sanity check asap

Reply:
decently optimistic that I found my bug: `fetch_and_update_state` was getting called with the orig seqno and so was noop-ing every time. if CI is happy, then I'll kick off a new nightlies run. I also owe a unit-level regression test for the bug I just hit, and probably we want to think a bit on if there's a way to restructure the new loop to avoid this category of bug

Reply:
tried a few different ways of writing a targeted unit-level regression test, but didn't manage to tickle the bug. given that even if I do manage something, it'll probably be brittle, and also that like every job in CI immediately broke, I might have to bail on that